#0: Smaller crt0 and better gp-relative addressing

nathan-TT opened the pull request:
In moving towards ExecuteInPlace kernel support, this is a small tweak to startup code and linker scripts.

1) Initialize GP using absolute, rather than pc-relative, addressing (The gp value is fixed, the pc values are not)

2) Allow relaxation on the following initializations, as gp has been initialized. This means that the `__ldm_bss_start` and `__ldm_bss_end` computations become a simple addition from gp.

3) There's a problem during relaxation to do with potential realignment. This means the linker cannot relax a gp-relative address of `gp - 0x800`, but has to add 4 bytes of fudge factor. It seems hard to get that optimal, and thus the linker script changes to initialize `__global_pointer$` to 0x7fc above `DATA_REGION`.  That allows the above relaxation. It's no loss in practice as the DATA_REGION is usually (always?) much smaller than 4K.

In summary, startup code goes from:
```
00000760 <_start>:
 760:   ffb00197                auipc   gp, 0xffb00
 764:   0a018193                addi    gp, gp, 160 # ffb00800 <__stack_top+0xfffff800>
 768:   ffb01117                auipc   sp, 0xffb01
 76c:   89810113                addi    sp, sp, -1896 # ffb01000 <__stack_top+0x0>
 770:   ffb00517                auipc   a0, 0xffb00
 774:   9f050513                addi    a0, a0, -1552 # ffb00160 <__stack_top+0xfffff160>
 778:   ffb00597                auipc   a1, 0xffb00
 77c:   f4458593                addi    a1, a1, -188 # ffb006bc <__stack_top+0xfffff6bc>
 780:   00000097                auipc   ra, 0x0
 784:   7f0080e7                jalr    2032(ra) # f70 <wzerorange>
 788:   ffb00917                auipc   s2, 0xffb00
 78c:   87890913                addi    s2, s2, -1928 # ffb00000 <__stack_top+0xfffff000>
 790:   ffb00997                auipc   s3, 0xffb00
 794:   87098993                addi    s3, s3, -1936 # ffb00000 <__stack_top+0xfffff000>
 798:   0100006f                j       7a8 <.L2^B1>
 ...
```
to:
```
00000760 <_start>:
 760:   ffb001b7                lui     gp, 0xffb00
 764:   7fc18193                addi    gp, gp, 2044 # ffb007fc <__stack_top+0xfffff7fc>
 768:   ffb01117                auipc   sp, 0xffb01
 76c:   89810113                addi    sp, sp, -1896 # ffb01000 <__stack_top+0x0>
 770:   96418513                addi    a0, gp, -1692 # ffb00160 <my_y>
 774:   ec018593                addi    a1, gp, -320 # ffb006bc <__ldm_bss_end>
 778:   7e8000ef                jal     ra, f60 <wzerorange>
 77c:   80418913                addi    s2, gp, -2044 # ffb00000 <__ldm_start>
 780:   80418993                addi    s3, gp, -2044 # ffb00000 <__ldm_start>
 784:   0100006f                j       794 <.L2^B1>
 ...
```
### Ticket
None created, noticed when working on XIP solutions.

### Problem description
Startup code larger than necessary.

### What's changed
* Allow post-gp initialization relaxation
* Adjust gp value to account for alignment factor.

### Checklist
- [YES] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

nathan-TT committed:
#0: Smaller crt0 and better gp-relative addressing

In moving towards ExecuteInPlace kernel support, this is a small tweak
to startup code and linker scripts.

1) Initialize GP using absolute, rather than pc-relative, addressing
(The gp value is fixed, the pcvalues are not)

2) Allow relaxation on the following initializations, as gp has been
initialized. This means that the __ldm_bss_start and __ldm_bss_end
computations become a simple addition from gp.

3) There's a problem during relaxation to do with potential
realignment. This means the linker cannot relax a gp-relative address
of gp -0x800, but has to add 4 bytes of fudge factor. It seems hard to
get that optimal, and thus the linker script changes to initialize
__global_pointer$ to 0x7fc above DATA_REGION.  That allows the above
relaxation.

In summary, startup code goes from:
00000760 <_start>:
 760:   ffb00197                auipc   gp, 0xffb00
 764:   0a018193                addi    gp, gp, 160 # ffb00800 <__stack_top+0xfffff800>
 768:   ffb01117                auipc   sp, 0xffb01
 76c:   89810113                addi    sp, sp, -1896 # ffb01000 <__stack_top+0x0>
 770:   ffb00517                auipc   a0, 0xffb00
 774:   9f050513                addi    a0, a0, -1552 # ffb00160 <__stack_top+0xfffff160>
 778:   ffb00597                auipc   a1, 0xffb00
 77c:   f4458593                addi    a1, a1, -188 # ffb006bc <__stack_top+0xfffff6bc>
 780:   00000097                auipc   ra, 0x0
 784:   7f0080e7                jalr    2032(ra) # f70 <wzerorange>
 788:   ffb00917                auipc   s2, 0xffb00
 78c:   87890913                addi    s2, s2, -1928 # ffb00000 <__stack_top+0xfffff000>
 790:   ffb00997                auipc   s3, 0xffb00
 794:   87098993                addi    s3, s3, -1936 # ffb00000 <__stack_top+0xfffff000>
 798:   0100006f                j       7a8 <.L2^B1>
 ...

to:
00000760 <_start>:
 760:   ffb001b7                lui     gp, 0xffb00
 764:   7fc18193                addi    gp, gp, 2044 # ffb007fc <__stack_top+0xfffff7fc>
 768:   ffb01117                auipc   sp, 0xffb01
 76c:   89810113                addi    sp, sp, -1896 # ffb01000 <__stack_top+0x0>
 770:   96418513                addi    a0, gp, -1692 # ffb00160 <my_y>
 774:   ec018593                addi    a1, gp, -320 # ffb006bc <__ldm_bss_end>
 778:   7e8000ef                jal     ra, f60 <wzerorange>
 77c:   80418913                addi    s2, gp, -2044 # ffb00000 <__ldm_start>
 780:   80418993                addi    s3, gp, -2044 # ffb00000 <__ldm_start>
 784:   0100006f                j       794 <.L2^B1>
 ...

pgkeller reviewed (approved):
This raises a good point: the local memory size on BH ierisc/brisc/ncrisc is 8K which is not 100% addressable w/ offsets off of GP.  We need to be sure that the low address range is covered (and not, eg, the stack!). we should also have a test that accesses data across the full range of available addresses fyi @abhullar-tt

nathan-TT committed:
#0: relax further

experimentation shows more headroom on GP value is needed.
Fix another __global_pointer initialization

nathan-TT commented:
Experimentation shows that 0x7fc was still too optimistic, with 0x7f0 all the kernels I tried correctly used gp-rel addressing for that start of DATA_REGION.
I also discovered another __global_pointer$ initializstion (ncrisc_resume).

nathan-TT committed:
Merge branch 'main' into nsidwell/pic-push

nathan-TT committed:
Merge branch 'main' into nsidwell/pic-push
